pydocstringformatter/run.py

```diff
@@ -112,30 +113,101 @@ def get_enabled_formatters(self) -> dict[str, _formatting.Formatter]:
         return enabled
 
     def format_file_tokens(
-        self, tokens: list[tokenize.TokenInfo]
+        self, tokens: list[tokenize.TokenInfo], filename: Path
     ) -> tuple[list[tokenize.TokenInfo], bool]:
-        """Format a list of tokens."""
+        """Format a list of tokens.
+
+        tokens: List of tokens to format.
+        filename: Name of the file the tokens are from.
+
+        Returns:
+            A tuple containing [1] the formatted tokens in a list
+            and a boolean indicating if the tokens were changed.
+
+        Raises:
+            UnstableResultError::
+                If the formatters are not able to get to a stable result.
+                It reports what formatters are still modifying the tokens.
+        """
         formatted_tokens: list[tokenize.TokenInfo] = []
         is_changed = False
 
         for index, tokeninfo in enumerate(tokens):
             new_tokeninfo = tokeninfo
 
             if _utils.is_docstring(new_tokeninfo, tokens[index - 1]):
-                for _, formatter in self.enabled_formatters.items():
-                    new_tokeninfo = formatter.treat_token(new_tokeninfo)
-            formatted_tokens.append(new_tokeninfo)
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo
+                )
+                is_changed = is_changed or bool(changers)
+
+                # Run formatters again to check if the result is stable
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo,
+                )
+
+                if not stable:
+                    conflicting_formatters = {
+                        k: v
+                        for k, v in self.enabled_formatters.items()
+                        if k in changers
+                    }
+                    template = _utils.create_gh_issue_template(
+                        new_tokeninfo, conflicting_formatters, str(filename)
+                    )
+
+                    raise UnstableResultError(template)
 
-            if tokeninfo != new_tokeninfo:
-                is_changed = True
+            formatted_tokens.append(new_tokeninfo)
 
         return formatted_tokens, is_changed
 
+    def apply_formatters_twice(
+        self, token: tokenize.TokenInfo
+    ) -> tuple[tokenize.TokenInfo, list[str], bool]:
+        """Apply the formatters twice to a token.
+
+        Also tracks which formatters changed the token.
+
+        Returns:
+            A tuple containing:
+            [1] the formatted token and
+            [2] a list of formatters that changed the token.
+            [3] a boolean indicating if the token was in the second pass.
+        """
+        stable = False
+        token, changers = self.apply_formatters(token)
+        if changers:
```

Comment:
If on the 3rd run there are `changers` won't we start a 4th run here?

Reply:
Yeah, but my question is regarding your previous comment where you said that this functions makes sure not to apply formatters a fourth time. I think this line makes it so it is not true.
If run 1 has changers, run 2 will be triggered by this check.
If run 2 has changers, run 3 will be triggered in the calling function.
If run 3 has changers, then this check starts a run 4. I think we want to avoid that.

Reply:
I think I follow changed it so it only applies once after the first double change